File: arrays_and_strings/q1.py
import sys

def all_unique(inputstr):
    sorted_inputstr = sorted(inputstr)
    for i, c in enumerate(sorted_inputstr[1:]):
        if c == sorted_inputstr[i]:
            print("FALSE: repeated character: %s" % c)
            return False

    print("TRUE: all unique characters in : %s" % inputstr)
    return True


if __name__ == "__main__":

    print("\nQ1: Implement an algo to determine if a string has all unique characters.\n\n")

    inputstrs = []

    if len(sys.argv) > 2:
        print("usage: python3 %s <string>" % sys.argv[0])
        sys.exit(-1)
    elif len(sys.argv) == 2:
        inputstrs.append(sys.argv[1])
    else:
        inputstrs = ["leolabs", "test", "dog"]

    for inputstr in inputstrs:
       all_unique(inputstr)

File: arrays_and_strings/q1.go
// Implement an algo to determine if a string has all unique characters.

package main

import (
	"fmt"
	"strings"
	"github.com/thiagorobert/interviewprep/thiutil"
)

func check_unique_chars(text string) {
	seen := make(map[byte]bool)
	for _, ch := range []byte(text) {
		if _, exists := seen[ch]; exists {
			fmt.Printf("%s contains repeated characters\n", text)
			break
		}
		seen[ch] = true
	}
	if len(seen) == len(text) {
		fmt.Printf("%s contains all unique characters\n", text)
	}
}

func bsortcheck_unique_chars(text string) {
	bytes := []byte(text)

	thiutil.Bsort(bytes)
	fmt.Printf("bubblesort: %s\n", string(bytes))

	has_repeated := false
	last_seen := bytes[0]
	for i := 1; i < len(bytes); i++ {
		if bytes[i] == last_seen {
			has_repeated = true
			break
		}
		last_seen = bytes[i]
	}

	if has_repeated {
		fmt.Printf("%s contains repeated characters\n", text)
	} else {
		fmt.Printf("%s contains all unique characters\n", text)
	}
}

func main() {
	r := thiutil.NewReader()

	for {
		text := r.In()
		check_unique_chars(text)
		bsortcheck_unique_chars(text)

		if strings.Compare("hi", text) == 0 {
			fmt.Println("hello, Yourself")
		}

		if strings.Compare("exit", text) == 0 {
			fmt.Println("exiting...")
			break
		}
	}
}
